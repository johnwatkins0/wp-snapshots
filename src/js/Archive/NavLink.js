/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import get from 'lodash.get';

const getLargeOrMediumImageUrl = imageSizes =>
  get(imageSizes, ['large', 'source_url'], get(imageSizes, ['medium', 'source_url'], ''));

const NavLink = ({
  index,
  title: { rendered: title },
  excerpt: { rendered: excerpt },
  image: {
    media_details: { sizes: imageSizes },
  },
  slug,
}) => (
  <div className="SnapshotsNavLink" style={{ animationDelay: `.${index}s` }}>
    <Link to={`/snapshot/${slug}/`} href={`/snapshot/${slug}/`} className="SnapshotsNavLink__inner">
      <div
        style={{ backgroundImage: `url('${getLargeOrMediumImageUrl(imageSizes)}')` }}
        className="SnapshotsNavLink__front"
      >
        <h2 dangerouslySetInnerHTML={{ __html: title }} className="SnapshotsNavLink__title" />
      </div>
      <div className="SnapshotsNavLink__back">
        <p dangerouslySetInnerHTML={{ __html: excerpt }} />
      </div>
    </Link>
  </div>
);

NavLink.propTypes = {
  index: PropTypes.number.isRequired,
  title: PropTypes.objectOf(PropTypes.string).isRequired,
  image: PropTypes.objectOf(PropTypes.any).isRequired,
  slug: PropTypes.string.isRequired,
  excerpt: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default NavLink;
