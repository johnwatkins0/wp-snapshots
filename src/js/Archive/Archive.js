/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';

import NavLink from './NavLink';
import { Facebook, Twitter } from '../svg';

const Archive = ({ posts, termName, termDescription }) => (
  <nav className="SnapshotsArchive">
    <div className="SnapshotsArchive__intro">
      <h1 className="SnapshotsArchive__title" dangerouslySetInnerHTML={{ __html: termName }} />
      <p
        className="SnapshotsArchive__description"
        dangerouslySetInnerHTML={{ __html: termDescription }}
      />
      <div className="SnapshotsArchive__social">
        <a
          target="_blank"
          href={`https://www.facebook.com/sharer/sharer.php?u=${window.location.href}`}
        >
          <Facebook />
        </a>
        <a
          target="_blank"
          href={`https://twitter.com/intent/tweet?url=${
            window.location.href
          }&text=${termDescription}`}
        >
          <Twitter />
        </a>
      </div>
    </div>
    {posts.map((post, index) => post.image && <NavLink key={post.id} index={index} {...post} />)}
  </nav>
);

Archive.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  termName: PropTypes.string.isRequired,
  termDescription: PropTypes.string.isRequired,
};

export default Archive;
