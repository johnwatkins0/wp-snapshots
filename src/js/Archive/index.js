import Archive from './Archive';

export { default as NavLink } from './NavLink';
export default Archive;
