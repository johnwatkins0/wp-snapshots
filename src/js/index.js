/* eslint react/no-danger: 0 */
import React, { Component } from 'react';
import { render } from 'react-dom';
import PropTypes from 'prop-types';
import { BrowserRouter, Route, Switch, Link } from 'react-router-dom';
import DocumentTitle from 'react-document-title';
import get from 'lodash.get';
import { parse } from 'qs';

import Archive from './Archive';
import Single from './Single';
import MiniNav from './MiniNav';

import { appDidMount } from './store/actions';
import effects from './store/effects';
import reducer from './store/reducer';
import { getPostBySlug, getPosts } from './store/selectors';

import { ROOT_SELECTOR } from './constants';
import { withScrollToTop } from './utils/withScrollToTop';
import { withReplaceEditLink } from './utils/withReplaceEditLink';

const ScrollToTopSingle = withScrollToTop(withReplaceEditLink(Single, 'post'));
const ScrollToTopArchive = withScrollToTop(withReplaceEditLink(Archive, 'category'));

const INITIAL_STATE = {
  posts: [],
};

export class App extends Component {
  static propTypes = {
    termId: PropTypes.string.isRequired,
    termName: PropTypes.string.isRequired,
    termSlug: PropTypes.string.isRequired,
    termDescription: PropTypes.string.isRequired,
    basenamePrefix: PropTypes.string,
    titleAppend: PropTypes.string,
  };

  static defaultProps = {
    basenamePrefix: '',
    titleAppend: '',
  };

  constructor(props) {
    super(props);

    this.state = INITIAL_STATE;
    this.dispatch = this.dispatch.bind(this);
    this.triggerEffect = this.triggerEffect.bind(this);
    this.getState = this.getState.bind(this);
  }

  componentDidMount() {
    this.dispatch(appDidMount(this.props.termId, this.props.previewPost || null));
  }

  getState = () => ({ ...this.state });

  dispatch(action) {
    this.setState(state => reducer(state, action), () => this.triggerEffect(action));
  }

  triggerEffect(action) {
    if (action.type in effects) {
      effects[action.type](action, {
        dispatch: this.dispatch,
        getState: this.getState,
      });
    }
  }

  render = () => (
    <BrowserRouter basename={this.props.basenamePrefix}>
      <main className="SnapshotsMain">
        <Switch>
          <Route
            path="/snapshots/:name/"
            render={() => (
              <DocumentTitle title={this.props.termName + this.props.titleAppend}>
                <ScrollToTopArchive
                  posts={getPosts(this.state)}
                  termName={this.props.termName}
                  termDescription={this.props.termDescription}
                  basename={this.props.basenamePrefix}
                  id={Number(this.props.termId)}
                />
              </DocumentTitle>
            )}
          />

          <Route
            path="/snapshot/:slug/"
            component={({
              match: {
                params: { slug },
              },
            }) => {
              const post = getPostBySlug(this.state, slug);
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

          <Route
            path="/"
            component={({ location }) => {
              const search = location.search || '';
              const params = parse(search.replace('?', ''));
              if (!('preview' in params) || !('p' in params)) {
                return null;
              }

              if (!this.props.previewPost) {
                return null;
              }

              const post = this.state.posts.find(({ id }) => id === Number(params.p));
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

        </Switch>
      </main>
    </BrowserRouter>
  );
}

export const startApp = () => {
  [...document.querySelectorAll(ROOT_SELECTOR)].forEach((root) => {
    render(
      <App
        termName={root.getAttribute('data-term-name')}
        titleAppend={root.getAttribute('data-title-append')}
        termId={root.getAttribute('data-term-id')}
        termSlug={root.getAttribute('data-term-slug')}
        termDescription={root.getAttribute('data-term-description')}
        basenamePrefix={root.getAttribute('data-basename-prefix').replace(/\/$/, '') || ''}
        previewPost={JSON.parse(root.getAttribute('data-preview-post'))}
      />,
      root,
    );
  });
};

export const start = async () => {
  if (!('fetch' in window)) {
    await import('whatwg-fetch');
  }

  startApp();
};

window.addEventListener('load', start);
