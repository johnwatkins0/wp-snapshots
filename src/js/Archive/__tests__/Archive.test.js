import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter, Route } from 'react-router-dom';

import { testPostsWithMedia } from '../../../../demo/src/data';

import Archive from '..';

Enzyme.configure({ adapter: new Adapter() });

const Component = () => (
  <BrowserRouter>
    <Route>
      <Archive termName="Name" termDescription="Description" posts={testPostsWithMedia} />
    </Route>
  </BrowserRouter>
);

describe('Archive', () => {
  it('renders', () => {
    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('has the correct number of items', () => {
    const wrapper = mount(<Component />);
    const navLinks = wrapper.find('.SnapshotsNavLink');
    expect(navLinks.length).toBe(7);
  });
});
