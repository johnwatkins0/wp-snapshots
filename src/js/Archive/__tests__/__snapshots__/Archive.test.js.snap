// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archive renders 1`] = `
<nav
  className="SnapshotsArchive"
>
  <div
    className="SnapshotsArchive__intro"
  >
    <h1
      className="SnapshotsArchive__title"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Name",
        }
      }
    />
    <p
      className="SnapshotsArchive__description"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Description",
        }
      }
    />
    <div
      className="SnapshotsArchive__social"
    >
      <a
        href="https://www.facebook.com/sharer/sharer.php?u=about:blank"
        target="_blank"
      >
        <svg
          height="24px"
          viewBox="0 0 1792 1792"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Facebook
          </title>
          <path
            d="M1376 128q119 0 203.5 84.5t84.5 203.5v960q0 119-84.5 203.5t-203.5 84.5h-188v-595h199l30-232h-229v-148q0-56 23.5-84t91.5-28l122-1v-207q-63-9-178-9-136 0-217.5 80t-81.5 226v171h-200v232h200v595h-532q-119 0-203.5-84.5t-84.5-203.5v-960q0-119 84.5-203.5t203.5-84.5h960z"
            fill="currentColor"
          />
        </svg>
      </a>
      <a
        href="https://twitter.com/intent/tweet?url=about:blank&text=Description"
        target="_blank"
      >
        <svg
          height="24px"
          viewBox="0 0 1792 1792"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Twitter
          </title>
          <path
            d="M1684 408q-67 98-162 167 1 14 1 42 0 130-38 259.5t-115.5 248.5-184.5 210.5-258 146-323 54.5q-271 0-496-145 35 4 78 4 225 0 401-138-105-2-188-64.5t-114-159.5q33 5 61 5 43 0 85-11-112-23-185.5-111.5t-73.5-205.5v-4q68 38 146 41-66-44-105-115t-39-154q0-88 44-163 121 149 294.5 238.5t371.5 99.5q-8-38-8-74 0-134 94.5-228.5t228.5-94.5q140 0 236 102 109-21 205-78-37 115-142 178 93-10 186-50z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".0s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/donec-vel-volutpat/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-7.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Donec vel volutpat",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Nunc quis libero</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".1s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/donec-ut-eros/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-6.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Donec ut eros",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Phasellus orci magna</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".2s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/vivamus-ullamcorper/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-5.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Vivamus ullamcorper",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Nulla facilisis</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".3s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/vestibulum-dapibus/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-4.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Vestibulum dapibus",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Nam gravida malesuada</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".4s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/ut-eget-orci/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-3.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Ut eget orci",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Integer cursus </p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".5s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/praesent-sollicitudin/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-2.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Praesent sollicitudin",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Curabitur mauris urna</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
  <div
    className="SnapshotsNavLink"
    style={
      Object {
        "animationDelay": ".6s",
      }
    }
  >
    <a
      className="SnapshotsNavLink__inner"
      href="/snapshot/nullam-ultricies/"
      onClick={[Function]}
    >
      <div
        className="SnapshotsNavLink__front"
        style={
          Object {
            "backgroundImage": "url('/demo/src/images/1024x683-1.jpg')",
          }
        }
      >
        <h2
          className="SnapshotsNavLink__title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Nullam ultricies",
            }
          }
        />
      </div>
      <div
        className="SnapshotsNavLink__back"
      >
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Sed at efficitur</p>
",
            }
          }
        />
      </div>
    </a>
  </div>
</nav>
`;
